src/wasm/wasm-type.cpp

```diff
@@ -1277,27 +1280,93 @@ void Canonicalizer::scanType(Type* type) {
   }
 }
 
-void Canonicalizer::makeReachabilityFixedPoint() {
+void Canonicalizer::makeAncestorsFixedPoint() {
   // Naively calculate the transitive closure of the reachability graph.
   bool changed;
   do {
     changed = false;
-    for (auto& entry : reaches) {
-      auto& reachable = entry.second;
-      std::unordered_set<TypeID> nextReachable;
-      for (auto& other : reachable) {
-        auto& otherReaches = reaches[other];
-        nextReachable.insert(otherReaches.begin(), otherReaches.end());
+    for (auto& entry : ancestors) {
+      auto& succs = entry.second;
+      std::unordered_set<TypeID> nextAncestors;
+      for (auto& other : succs) {
+        auto& otherAncestors = ancestors[other];
+        nextAncestors.insert(otherAncestors.begin(), otherAncestors.end());
       }
-      size_t oldSize = reachable.size();
-      reachable.insert(nextReachable.begin(), nextReachable.end());
-      if (reachable.size() != oldSize) {
+      size_t oldSize = succs.size();
+      succs.insert(nextAncestors.begin(), nextAncestors.end());
+      if (succs.size() != oldSize) {
         changed = true;
       }
     }
   } while (changed);
 }
 
+std::vector<Canonicalizer::Item> Canonicalizer::getOrderedItems() {
+  // Topologically sort the Types and HeapTypes so that all children are
+  // canonicalized before their parents.
+
+  std::vector<TypeID> sorted;
+  std::unordered_set<TypeID> seen;
+
+  // Topologically sort so that all parents are pushed before their children.
+  // This sort will be reversed later to have children before their parents.
+  std::function<void(TypeID)> visit = [&](TypeID i) {
+    if (seen.count(i)) {
+      return;
+    }
+    // Push ancestors of the current type before pushing the current type.
+    auto it = ancestors.find(i);
+    if (it != ancestors.end()) {
+      for (auto ancestor : it->second) {
+        visit(ancestor);
+      }
+    }
+    seen.insert(i);
+    sorted.push_back(i);
+  };
+
+  // Collect the items to be sorted, including all the result HeapTypes and any
+  // type that participates in the ancestor relation.
+  std::set<TypeID> allIDs;
```

Comment:
```suggestion
  std::unordered_set<TypeID> allIDs;
```
I doubt it matters for speed, but it may be clearer to read, as it indicates there is no sorting order in this data structure?

Reply:
I intentionally used a `set` here so that the result of the topological would be deterministic. I'll add a comment.

Reply:
Ah, then I missed something.

Do you mean it keeps the sort at each topological depth deterministic?

Reply:
I meant for this `std::set` to keep the top-level driving of the sort deterministic and for the `std::set` in `ancestors` to keep each recursive step of the sort deterministic. But now that I think about it, the TypeIDs themselves aren't deterministic, so using a `std::set` here is not sufficient to make the sort deterministic. Furthermore, it shouldn't matter whether this sort is deterministic; that shouldn't be visible to the user or affect anything else. I'll switch to using `std::unordered_set` here and in `ancestors`.

Reply:
Thanks! Sounds good.